---
title: "AsaruSim data analysis"
subtitle: "Comparaison real vs simulation"
author: "Ali Hamraoui"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: hide
    code_download: true
    toc: true
    toc_float: true
    number_sections: false
---

<style>
body {
text-align: justify}
</style>

<!-- Automatically computes and prints in the output the running time for any code chunk -->
```{r, echo=FALSE}
# https://github.com/rstudio/rmarkdown/issues/1453
hooks = knitr::knit_hooks$get()
hook_foldable = function(type) {
  force(type)
  function(x, options) {
    res = hooks[[type]](x, options)
    
    if (isFALSE(options[[paste0("fold_", type)]])) return(res)
    
    paste0(
      "<details><summary>", "show", "</summary>\n\n",
      res,
      "\n\n</details>"
    )
  }
}
knitr::knit_hooks$set(
  output = hook_foldable("output"),
  plot = hook_foldable("plot")
)
```

<!-- Set default parameters for all chunks -->
```{r, setup, include = FALSE}
set.seed(1)
knitr::opts_chunk$set(echo = TRUE, # display code
                      # display chunk output
                      message = FALSE,
                      warning = FALSE,
                      fold_output = FALSE, # for sessionInfo()
                      fold_plot = FALSE,
                      
                      # figure settings
                      fig.align = 'center',
                      fig.width = 20,
                      fig.height = 15)
```

## R Markdown
```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
library(dplyr)
library(ggplot2)
library('EnsDb.Hsapiens.v79')
library("SingleR")
library(Seurat)
options(Seurat.object.assay.version = "v5")
source("./imports.R")
```


### 5k Human PBMCs data
IN this section, we downloaded the [5k Human PBMCs, 3â€™ v3.1](https://www.10xgenomics.com/datasets/5k-human-pbmcs-3-v3-1-chromium-controller-3-1-standard) count matrix and used it as input to AsaruSim. From the 5,000 cells initially present in the original matrix, we selected 3 cell types (CD8+T, CD4+T and B cells) resulting in 1,091 cells.

We load the 10X matrix
```{r load_PBMC_data}
pbmc.mtx <- Seurat::Read10X("data/PBMC_10X/")
dim(pbmc.mtx)
sum(pbmc.mtx)
```

We create a Seurat object, identified clsters and cell types.
```{r cluster_PBMC_data, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
whitelist_cells <- gsub("-1$", "", colnames(pbmc.mtx))
colnames(pbmc.mtx) <- whitelist_cells

pbmc.Obj <- Seurat::CreateSeuratObject(pbmc.mtx, project = "PBMC", min.cells = 3, min.features =200)
pbmc.Obj <- clustuerer(pbmc.Obj, 200, 2000, 15)
pbmc.Obj <- auto.annotation(mtx=pbmc.mtx, obj=pbmc.Obj)
```

We visualize clusters and cell types
```{r umap_PBMC, fig.height=5, fig.width=12}
pbmc.Obj
Seurat::DimPlot(pbmc.Obj, reduction = "umap", label = T) | Seurat::DimPlot(pbmc.Obj, reduction = "umap", group.by = "cell_type", label = T)
```

We identified B-cells, CD4+ T-cells and CD8+ T-cells clusters.
```{r cell_selection}
df.identities <- cbind(as.data.frame(pbmc.Obj$cell_type),
                       as.data.frame(pbmc.Obj$seurat_clusters))
colnames(df.identities) <- c("cell_type","cluster")

list.identities <- list(
B_cells = which(df.identities$cluster %in% c(4) & 
                  df.identities$cell_type=="B-cells"),

T_cells = which( df.identities$cluster %in% c(5) & 
                   df.identities$cell_type=="CD4+ T-cells"),

macrophages = which(df.identities$cluster %in% c(3) & 
                      df.identities$cell_type=="CD8+ T-cells"))

selected.cells <- as.vector(unlist(list.identities))
selected.cells <- rownames(df.identities)[selected.cells]
```

We filter for selected cell types.
```{r filter_Obj}
pbmc.Obj[["CellName"]] <- colnames(pbmc.Obj)
sub.pbmc.Obj <- subset(pbmc.Obj, subset = CellName %in% selected.cells)
```

We re-compute clusters and visualize them.
```{r viz_sub_data, fig.height=5, fig.width=12}
sub.pbmc.Obj <- clustuerer(sub.pbmc.Obj, 200, 2000, 15)
Seurat::DimPlot(sub.pbmc.Obj, reduction = "tsne", label = T) | Seurat::DimPlot(sub.pbmc.Obj, reduction = "tsne", group.by = "cell_type", label = T) 
```
We save the matrix
```{r create_sub_matrix}
sum(pbmc.mtx[, selected.cells])
pbmc.mtx_sub <- pbmc.mtx[, selected.cells]
dim(pbmc.mtx_sub)
sum(pbmc.mtx_sub)
write.csv(pbmc.mtx_sub, file="data/pbmc.mtx_sub_10M.csv")
```

### Comparaison: AsaruSim simulated PBMC vs real PBMC

In this section used the subset of the PBMC matrix as a template to simulate the synthetic UMI count matrix (step1). Next, we simulated 30 million perfect reads (FASTA) (step2) with 2 PCR cycles (step 3). We downloaded a subset of 1M original [FASTQ raw reads](https://www.10xgenomics.com/datasets/5k-human-pbmcs-3-v3-1-chromium-controller-3-1-standard) to generate the error model for Badread, and then introduced errors to generate the synthetic reads (FASTQ) (step4).

Next, we pre-processed the simulated raw reads using Sockeye pipeline (nanoporetech, 2023), then, both matrices were loaded and processed using Seurat v5 (Hao et al. 2024). 

We reload the subseted PBBMC matrix
```{r load_sub_matrix}
pbmc.mtx_sub <- read.csv("data/pbmc.mtx_sub_10M.csv", row.names = 1)
dim(pbmc.mtx_sub)
sum(pbmc.mtx_sub)
```

We load AsaruSim Marix preprocessed using Sockeye. 
```{r load_sim_matrix}
sim.mtx <- read.csv("data/Sockeye_PBMC_3/gene_expression.counts_5prim.tsv", sep = "\t", row.names = 1)
dim(sim.mtx)
sum(sim.mtx)
```

```{r}
sim.mtx <- rename_mtx_with_gene_names(sim.mtx)
```

We create a Seaurate object for the both datasets.
```{r clusters_and_annotation, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
real.Obj <- Seurat::CreateSeuratObject(pbmc.mtx_sub, project = "Real_PBMC")
real.Obj <- clustuerer(real.Obj, 200, 2000, 15)
real.Obj <- auto.annotation(mtx=pbmc.mtx_sub, obj=real.Obj)

sim.Obj <- Seurat::CreateSeuratObject(sim.mtx, project = "Sim_PBMC")
sim.Obj <- clustuerer(sim.Obj, 200, 2000, 15)
sim.Obj <- auto.annotation(mtx=sim.mtx, obj=sim.Obj)
```

```{r}
real.Obj
sim.Obj
```

We compared the TSNE visualisation of the both dataset.
```{r tsne_comparaison, fig.width=10, message=FALSE, warning=FALSE, paged.print=FALSE, fig.height=5}
real.plt <- Seurat::DimPlot(real.Obj, reduction = "tsne",  group.by = "corrected_cell_type", cols = c('#E377C2', '#BCBD22','#17BECF' )) + 
    ggtitle(paste('Real PBMC'))+ 
    theme(aspect.ratio = 1)

sim.plt <- Seurat::DimPlot(sim.Obj, reduction = "tsne",  group.by = "corrected_cell_type", cols = c('#E377C2', '#BCBD22','#17BECF' )) + 
    ggtitle(paste('Simuated PBMC'))+ 
    theme(aspect.ratio = 1)
real.plt | sim.plt
```

```{r}
B.real.markers <- FindMarkers(real.Obj, ident.1 = "B-cells", group.by = "cell_type")
CD4.real.markers <- FindMarkers(real.Obj, ident.1 = "CD4+ T-cells", group.by = "cell_type") 
CD8.real.markers <- FindMarkers(real.Obj, ident.1 = "CD8+ T-cells", group.by = "cell_type") 

B.sim.markers <- FindMarkers(sim.Obj, ident.1 = "B-cells", group.by = "cell_type")
CD4.sim.markers <- FindMarkers(sim.Obj, ident.1 = "CD4+ T-cells", group.by = "cell_type") 
CD8.sim.markers <- FindMarkers(sim.Obj, ident.1 = "CD8+ T-cells", group.by = "cell_type") 
```

```{r fig.height=5, fig.width=5}
B.markers <- merge(B.sim.markers, B.real.markers, by = 'row.names', all = F)
CD4.markers <- merge(CD4.sim.markers, CD4.real.markers, by = 'row.names', all = F)
CD8.markers <- merge(CD8.sim.markers, CD8.real.markers, by = 'row.names', all = F)

B.markers$cell_type <- 'B-cells'
CD4.markers$cell_type <- 'CD4+ cells'
CD8.markers$cell_type <- 'CD8+ cells'

markers <- rbind(B.markers, CD4.markers, CD8.markers)
```


```{r fig.height=6, fig.width=8}
ggplot(markers, aes(x=abs(avg_log2FC.x), y=abs(avg_log2FC.y)) ) +
  geom_bin2d(bins = 150, drop= T) +
  scale_fill_continuous(type = "viridis") +
  theme_bw()+
  ggtitle(paste0('Spearman R = ', round(cor(markers$avg_log2FC.x, 
              markers$avg_log2FC.y), 2)
              ))
```

#### Integration.

We integrate the both dataset and calculate a miLISI score.
```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
sim.Obj <- Seurat::CreateSeuratObject(sim.mtx, project = "Simulation")
real.Obj <- Seurat::CreateSeuratObject(pbmc.mtx_sub, project = "Real_PBMC")
integrated.Sobj <- integrate(sim.Obj, real.Obj)
lisi <- calc_meanLISI(integrated.Sobj, "umap")
```

The integration of the both dataset shows a miLISI value of 1.58.
```{r fig.height=6, fig.width=8}
DimPlot.integ(integrated.Sobj, "tsne")
```



